includes/admin/class-wc-rest-stripe-tokens-controller.php

```diff
@@ -0,0 +1,64 @@
+<?php
+/***
+ * Class WC_REST_Stripe_Tokens_Controller
+ */
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * REST controller for tokens.
+ */
+class WC_REST_Stripe_Tokens_Controller extends WC_Stripe_REST_Base_Controller {
+
+	/**
+	 * Endpoint path.
+	 */
+	protected $rest_base = 'wc_stripe/tokens';
+
+	/**
+	 * Register REST API routes for Stripe tokens.
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			// For more info on Stripe tokens, see the following:
+			// https://stripe.com/docs/api/tokens/object
+			'/' . $this->rest_base . '/(?P<token_id>[a-z]{3}_[a-zA-Z0-9]{24})',
+			[
+				'methods'             => WP_REST_Server::READABLE,
+				'callback'            => [ $this, 'get_token' ],
+				'permission_callback' => [ $this, 'check_permission' ],
+			]
+		);
+	}
+
+	/**
+	 * Retrieve a Stripe token, given a secret-key and token_id.
+	 *
+	 * @param WP_REST_Request $request Request object.
+	 *
+	 * @return WP_REST_Response Response object.
+	 */
+	public function get_token( $request ) {
+		$token_id   = $request->get_param( 'token_id' );
+		$secret_key = $request->get_header( 'X-WCStripe-Secret-Key' );
+
+		// Ensure secret_key is validated.
+		if ( ! preg_match( '/^\w+$/', $secret_key ) ) {
+			return new WP_REST_Response( [ 'result' => 'bad_request' ], 400 );
+		}
```

Comment:
I agree with the idea of removing some repetition. But in this case, if a `secret_key` with invalid syntax is given, do we want to add overhead by pinging Stripe's server, or do we want to short-circuit the call?

I don't think it will be common that invalid syntax will be input. So, I'll go ahead and remove this validation and let the try/catch block handle it 🙂

Reply:
> do we want to add overhead by pinging Stripe's server, or do we want to short-circuit the call?

I guess it would depend on how Stripe handles invalid keys, and if invalid keys with the wrong format are any different. I guess we could simply rely on Stripe's response, as it's the only source of truth.